File: tests/samples/Union.scala
package org.scalanative.bindgen.samples

import scala.scalanative.unsigned._
import scala.scalanative.unsafe._

@link("bindgentests")
@extern
object Union {
  type enum_union_op = CUnsignedInt
  object enum_union_op {
    final val UNION_SET: enum_union_op = 0.toUInt
    final val UNION_TEST: enum_union_op = 1.toUInt
  }

  type struct_s = CStruct1[CInt]
  type union_values = CArray[Byte, Nat._8]
  def union_get_sizeof(): CInt = extern
  def union_test_int(v: Ptr[union_values], op: enum_union_op, value: CInt): CInt = extern
  def union_test_long(v: Ptr[union_values], op: enum_union_op, value: CLong): CInt = extern
  def union_test_long_long(v: Ptr[union_values], op: enum_union_op, value: CLongLong): CInt = extern
  def union_test_double(v: Ptr[union_values], op: enum_union_op, value: CDouble): CInt = extern
  def union_test_string(v: Ptr[union_values], op: enum_union_op, value: CString): CInt = extern
  def union_test_struct(v: Ptr[union_values], op: enum_union_op, value: Ptr[struct_s]): CInt = extern

  object implicits {
    implicit class struct_s_ops(val p: Ptr[struct_s]) extends AnyVal {
      def a: CInt = p._1
      def a_=(value: CInt): Unit = p._1 = value
    }

    implicit class union_values_pos(val p: Ptr[union_values]) extends AnyVal {
      def l: Ptr[CLong] = p.asInstanceOf[Ptr[CLong]]
      def l_=(value: CLong): Unit = !p.asInstanceOf[Ptr[CLong]] = value
      def i: Ptr[CInt] = p.asInstanceOf[Ptr[CInt]]
      def i_=(value: CInt): Unit = !p.asInstanceOf[Ptr[CInt]] = value
      def ll: Ptr[CLongLong] = p.asInstanceOf[Ptr[CLongLong]]
      def ll_=(value: CLongLong): Unit = !p.asInstanceOf[Ptr[CLongLong]] = value
      def d: Ptr[CDouble] = p.asInstanceOf[Ptr[CDouble]]
      def d_=(value: CDouble): Unit = !p.asInstanceOf[Ptr[CDouble]] = value
      def s: Ptr[CString] = p.asInstanceOf[Ptr[CString]]
      def s_=(value: CString): Unit = !p.asInstanceOf[Ptr[CString]] = value
      def structInUnion: Ptr[struct_s] = p.asInstanceOf[Ptr[struct_s]]
      def structInUnion_=(value: Ptr[struct_s]): Unit = !p.asInstanceOf[Ptr[struct_s]] = value
    }
  }

  object struct_s {
    import implicits._
    def apply()(implicit z: Zone): Ptr[struct_s] = alloc[struct_s]
    def apply(a: CInt)(implicit z: Zone): Ptr[struct_s] = {
      val ptr = alloc[struct_s]
      ptr.a = a
      ptr
    }
  }
}

File: tests/samples/Union.c
#include "Union.h"
#include <stdlib.h>
#include <string.h>

int union_get_sizeof() { return sizeof(union values); }

int union_test_int(union values *v, enum union_op op, int value) {
    switch (op) {
    case UNION_SET:
        v->i = value;
        return 1;
    case UNION_TEST:
        return v->i == value;
    }
}

int union_test_long(union values *v, enum union_op op, long value) {
    switch (op) {
    case UNION_SET:
        v->l = value;
        return 1;
    case UNION_TEST:
        return v->l == value;
    }
}

int union_test_long_long(union values *v, enum union_op op, long long value) {
    switch (op) {
    case UNION_SET:
        v->ll = value;
        return 1;
    case UNION_TEST:
        return v->ll == value;
    }
}

int union_test_double(union values *v, enum union_op op, double value) {
    switch (op) {
    case UNION_SET:
        v->d = value;
        return 1;
    case UNION_TEST:
        return v->d == value;
    }
}

int union_test_string(union values *v, enum union_op op, const char *value) {
    switch (op) {
    case UNION_SET:
        v->s = value;
        return 1;
    case UNION_TEST:
        return v->s == value || !strcmp(v->s, value);
    }
}

int union_test_struct(union values *v, enum union_op op, struct s *value) {
    switch (op) {
    case UNION_SET:
        v->structInUnion.a = value->a;
        return 1;
    case UNION_TEST:
        return v->structInUnion.a == value->a;
    }
}
